Initial documentation for Aurora debugging (#645)
argonne-lcf/user-guides: docs/aurora/debugging/ddt-aurora.md

# Debugging on Aurora with DDT

We have licenses for Linaro DDT on Aurora, a parallel debugger that is able to use [gdb-oneapi](./gdb-oneapi.md) as its underlying engine on Aurora. This is not a tutorial on DDT; for that you should use Linaro's documentation, such as the documentation bundled with their client programs. Here we provide specific information on using DDT interactively in the GUI client-server mode on Aurora.

> ##Client

Reviewer: ```suggestion
## Client
```

Download and install the latest Linaro Forge client for your desktop/laptop system from the [Linaro website](https://www.linaroforge.com/download-documentation/). This is available for linux, MacOS, and Windows systems.

> ###Configuring the Remote Client

Reviewer: ```suggestion
### Configuring the Remote Client
```

Before you can start a DDT debugging session on Aurora compute nodes, you must set your client for remote connection from Aurora compute nodes. Your client window should look something like this:

![Example DDT client screen](ddt_client.png "Example DDT client screen")

Click the Remote Launch pull-down and click Configure to create a connector for Aurora:

![Click Configure](ddt_configure.png "Click Configure")

Click the Add button on the Configure Remote Connections screen:

![Click Add](ddt_configure_add.png "Click Add")

Create a configuration named "aurora," and set it up like this example, replacing "username" with your actual ALCF login name:

![Configuration example for aurora](ddt_configure_aurora.png "Configuration example for aurora")

You may want to test the configuration. To do that, click the Test Remote Launch button. If you see a login prompt like the following example, use your usual ALCF one-time password:

![Example DDT remote connection login prompt](ddt_login_prompt.png "Example DDT remote connection login prompt")

If the test is successful, you are are ready to proceed from an Aurora compute node.

> ###Invoking the DDT Server from Aurora

Reviewer: same as above and below

Author: I put in the missing space before "Invoking" by hand.

To run DDT interactively from Aurora, start up an [interactive PBS job](../running-jobs-aurora.md#interactive-jobs-on-compute-nodes). You'll need to load a module to access DDT:

```
module load forge
```

If you are using a wrapper script to map MPI ranks to PVC GPU tiles, you must set this environment variable to the full path to that wrapper script:

```
export FORGE_DEBUGGER_WRAPPER=/opt/aurora/24.180.1/support/tools/mpi_wrapper_utils/gpu_tile_compact.sh
```

> As discussed w.r.t. [gdb-oneapi](./gdb-oneapi.md), you must explicitly enable gpu debugging on all the PVC GPUs you are using, on all the nodes you are using. One way to do this is to create a script and execute it across all your compute nodes using `mpiexec`. Here is an example script, which takes an argument 1 to enable debugging or 0 to disable it:

Reviewer: ```suggestion
As discussed with respect to [gdb-oneapi](./gdb-oneapi.md), you must explicitly enable GPU debugging on all the PVC GPUs you are using, on all the nodes you are using. One way to do this is to create a script and execute it across all your compute nodes using `mpiexec`. Here is an example script, which takes an argument `1` to enable debugging or `0` to disable it:
```

> ```

Reviewer: ```suggestion
```bash linenums="1" title="helper_toggle_eu_debug.sh"
```

#!/usr/bin/env bash

> # helper_toggle_eu_debug.bash

Reviewer: ```suggestion
# helper_toggle_eu_debug.sh
```

export MY_RANK=${PMIX_RANK}
export MY_NODE=${PALS_NODEID}
export MY_LOCAL_RANK=${PALS_LOCAL_RANKID}

eu_debug_toggle() {
  for f in /sys/class/drm/card*/prelim_enable_eu_debug
  do
    echo $1 > $f
  done
  echo "INFO: EU debug state on rank-${MY_RANK}: $(cat /sys/class/drm/card*/prelim_enable_eu_debug | tr '\n' ' ')"
  # sleep 10
}

# One rank per node toggles eu debug:
if [ ${MY_LOCAL_RANK} -eq 0 ]; then
    eu_debug_toggle $1
    touch eu_debugs/eu_debug_toggled0_${MY_NODE}.flag
fi
```

From the interactive prompt on your lead Aurora compute node, issue

```
export NNODES=`wc -l < $PBS_NODEFILE`

> mpiexec -n $NNODES ./helper_toggle_eu_debug.bash 1

Reviewer: ```suggestion
mpiexec -n $NNODES ./helper_toggle_eu_debug.sh 1
```
